### PR TITLE
isso: module & init at 0.9.10 WIP

### PR DESCRIPTION
**Pull request by Profpatsch**
isso commenting server

not sure if the way I handle misaka 1&2 is the way to go @garbas 

cc @aszlig (spacing. :) )

### CODE REVIEW

**Comment by aszlig** [comment]
Feel free to add `Cc: @aszlig - nitpick` in future PRs if you want even more nitpicking (I could nitpick at even more here as well, for example on the missing commit message :goberserk:, but I'll stop now), the whitespace thing was just because it immediately sprang into my eyes :-D

**Comment by aszlig** [comment]
Cc: @edolstra because of the addition of a new function in `lib/attrsets.nix`.

File: lib/attrsets.nix [hunk 1]
```diff
@@ -43,6 +44,16 @@ rec {
       [ { name = head attrPath; value = setAttrByPath (tail attrPath) value; } ];
 
 
+  /* Map function f over the value of the attribute described by attrPath.
+     For instance, mapAttrByPath [ "x" "y" ] (a: a+1) { x.y = 1; } returns
+     { x.y = 2; }. If the path doesn’t exist the default value is used. */
```
Comment (edolstra) [inline lib/attrsets.nix]: It's not clear to be from this description what is being mapped. That suggests that multiple values are updated, but the example suggests only the specified attribute (x.y) is updated. And what is the `recursiveUpdate` for?

Comment (Profpatsch) [inline lib/attrsets.nix]: Changed the implementation in the newest version, according to aszlig’s input.

Do you think we should rename that function? I can’t think of another word but map. But I will change the docstring to start with “Apply function …”.

**Comment by Profpatsch** [comment]
I renamed the functions to `setAttrInSetByPath` and `updateAttrByPath` @edolstra.

I also added tests and fixed errors with certain inputs for the set function (gah, dynamically typed languages ;) ).

**Comment by davidak** [comment]
i would like to use isso on nixos, so i tested this PR:

https://github.com/davidak/nixos-config/blob/d1fcef16470c567245fcb480c55f980b59bfc208/binary/configuration.nix#L129

it got installed, configuration got created and the service started, but with errors.

```
Traceback (most recent call last):
  File "/nix/store/iihfxv3fi5i2z8arsr8b1kgp2q399x6f-python2.7-isso-0.9.10/bin/.isso-wrapped", line 12, in <module>
    sys.exit(main())
  File "/nix/store/iihfxv3fi5i2z8arsr8b1kgp2q399x6f-python2.7-isso-0.9.10/lib/python2.7/site-packages/isso/__init__.py", line 271, in main
    wsgi.SocketHTTPServer(sock, make_app(conf)).serve_forever()
  File "/nix/store/iihfxv3fi5i2z8arsr8b1kgp2q399x6f-python2.7-isso-0.9.10/lib/python2.7/site-packages/isso/wsgi.py", line 203, in __init__
    HTTPServer.__init__(self, sock, SocketWSGIRequestHandler)
  File "/nix/store/awbf3jh262m1k16hxgbgngdmxv66gd79-python-2.7.11/lib/python2.7/SocketServer.py", line 420, in __init__
    self.server_bind()
  File "/nix/store/awbf3jh262m1k16hxgbgngdmxv66gd79-python-2.7.11/lib/python2.7/BaseHTTPServer.py", line 110, in server_bind
    self.server_name = socket.getfqdn(host)
  File "/nix/store/awbf3jh262m1k16hxgbgngdmxv66gd79-python-2.7.11/lib/python2.7/socket.py", line 141, in getfqdn
    hostname, aliases, ipaddrs = gethostbyaddr(name)
TypeError: must be string without null bytes, not str
```
that might be related to https://github.com/NixOS/nixpkgs/issues/1248 but it also don't work with a workaround.

**Comment by Profpatsch** [comment]
@davidak Does [this](https://gist.github.com/Profpatsch/e27373dfe346dc4e59838860ba3982f8) work when you put it into a container, like so:

    containers.isso.config = ./isso-test.nix;

and then start it like

 ```
sudo systemctl restart container@isso
sudo journalctl -M isso
```

**Comment by davidak** [comment]
@Profpatsch yes.

bildschirmfoto 2016-07-17 um 16 42 59

(only some Umlauts are broken in issos interface)

**Comment by davidak** [comment]
The Umlauts are fine when i use it in [Nikola](https://getnikola.com/), maybe an encoding problem in your simple example website.

Your config works also without the container and then the comments are persistent and every restart works!

The difference of our configs is:

your working config:
```
      otherConfig = {
        server.listen = "http://localhost:9001";
      };
```

my try:
```
otherConfig = { server = { listen = "localhost:9090"; }; };
```

the error was the missing `http://`!

please change that in the example of the module:
`Example: { general = { log-file = "/var/lib/isso/isso.log"; } ; server = { listen = "localhost:8080"; } ; }`

my whole config: https://github.com/davidak/nixos-config/commit/c8b6443b90bf9e073f662ec14946acf84a176455

**Comment by Profpatsch** [comment]
> After a restart are the comments gone, also every second restart fails.

The example page doesn’t specify an encoding, as you can see in the code



> please change that in the example of the module:
> Example: { general = { log-file = "/var/lib/isso/isso.log"; } ; server = { listen = "localhost:8080"; } ; }

Hm, interesting. I used `http://` in the test as well. But it shouldn’t be necessary imo.

**Comment by davidak** [comment]
>But it shouldn’t be necessary imo.

have you tested it?

then resolve the merge conflicts and get it merged.

next step would be to use a WSGI Server like gunicorn for production use. the developer's choice is uWSGI but the debian package uses gunicorn. it is simpler to configure and has also a good performance. should be good most users.

https://posativ.org/isso/docs/extras/deployment/
https://github.com/jgraichen/debian-isso/blob/master/debian/isso.service
https://sicherheitskritisch.de/2015/04/sichere-installation-und-konfiguration-des-kommentarsystems-isso-auf-debian/

**Comment by Profpatsch** [comment]
> then resolve the merge conflicts and get it merged.

All in due time. It’s a work in progress, after all. Except if you’d like to help me. There’s a few tests to be written. :)

**Comment by davidak** [comment]
@Profpatsch How could i help? I haven't written module tests by now and don't know what you have in mind. So it might be more effective when you do that. Do you think you find some time in the near future?

**Comment by Profpatsch** [comment]
> I haven't written module tests by now and don't know what you have in mind.

I haven’t either. :)
What I want are some automatic VM tests that check if the service works from a very high level. As in checking the actual output of a test website, by posting a comment and then grepping the site output for the content of that comment. That way the maintainer is immediately notified if the service breaks.


> Do you think you find some time in the near future?

This Saturday/Sunday at the earliest.

**Comment by davidak** [comment]
- `general.dbpath` don't get added into the config, the database don't get created, all comments are lost after restarting the service
- import created `/tmp/isso.db` but the service don't loads it
- `general.notify` also don't land in the config
- the current version of Isso is `0.10.4`, this is `0.9.10`
- there is a newline after each config line:

```
[general]
host = davidak.binary.lan.davidak.de
 davidak.de


[guard]
direct-reply = 3

enabled = true

ratelimit = 2

reply-to-self = false


```

**Comment by matthiasbeyer** [comment]
I guess this has stalled.

**Comment by Profpatsch** [comment]
Yeah, it’s pretty low on my priority list right now.

**Comment by davidak** [comment]
@Profpatsch sadly i had also no time to look into this. i have planned to deploy my new website without comments first next month (what is tomorrow :)). so i can make progress and fix small things like that later.

**Comment by Profpatsch** [comment]
I actually planned on using isso as commenting server for audio files (soundcloud-like), but that’s a very far-off project right now.

**Pull request closed without merging**